app/components/chart-selection-tabs.tsx

```diff
@@ -277,16 +281,58 @@ const PublishChartButton = () => {
     }
   });
 
-  return (
+  const [editingPublishedChart, setEditingPublishedChart] =
+    React.useState(false);
+  const [loaded, setLoaded] = React.useState(false);
+  const { asPath } = useRouter();
+  const session = useSession();
+  const chartId = getRouterChartId(asPath);
+
+  React.useEffect(() => {
+    const run = async () => {
+      if (session.data?.user && chartId) {
+        const config = await fetchChartConfig(chartId);
+
+        if (config?.user_id === session.data.user.id) {
+          setEditingPublishedChart(true);
+        }
+      }
+
+      if (session.status !== "loading") {
+        setLoaded(true);
+      }
+    };
+
+    run();
+  }, [chartId, session]);
```

Comment:
* It seems like the editingPublishedChart could be computed.
* I think it would be a bit more readable if we could get rid of the effect and use something similar to useQuery from react-query

```
// Not tested (asked chatgpt but it looked good)
const useFetchData = ({ queryFn, enabled = true } = {}) => {
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);
  const [fetchStatus, setFetchStatus] = useState('idle'); // Possible values: 'idle', 'fetching', 'success', 'error'

  useEffect(() => {
    if (!enabled) {
      return;
    }

    const fetchData = async () => {
      setFetchStatus('fetching');
      try {
        const result = await queryFn();
        setData(result);
        setFetchStatus('success');
      } catch (err) {
        setError(err);
        setFetchStatus('error');
      }
    };

    fetchData();
  }, [enabled, queryFn]);

  return { data, error, fetchStatus };
};


// Then in your code
  const { asPath } = useRouter();
  const session = useSession();
  const { loaded, data } = useFetchData({ queryFn: fetchChartData, enabled: session.data.user && chartId })
  
    const editingPublishedChart = session.data.user.id && config?.user_id === session.data.user.id
```

I feel like using the same pattern as react-query makes it closer to declarative code, and easier to read, what do you think ?

Reply:
It would prevent also infinite loading when the fetch fails ;)

Reply:
Yup, good point! Thanks for taking a look and making a proposal, I'll adjust the code 👌